### PR TITLE
Restrict squeeze dims to prevent squeezing out batch dim

### PR DESCRIPTION
**Pull request by elibixby**
Unless `tf.squeeze` is given an explicit list of axis to squeeze which does not include the batch dimension, it will squeeze batch sizes of 1 into scalars, which will cause `sparse_softmax_cross_entropy_with_logits` to fail with a shape error. This only happens in training if a `allow_final_batches=True` is allowed to trigger, which must require a restriction not based on number of steps but based on epochs.

### CODE REVIEW

**Comment by elibixby** [comment]
Assigned to @puneith to merge when e2e tests are merged and we move off of Google3
